DIPs/13NNN-WGB.md

```diff
@@ -0,0 +1,893 @@
+parameter, indicates that the return is by Move Ref, and behaves equivalently
+to the non-EMO semantics of:
+
+```
+struct T { } // T is not an EMO
+
+ref T func(return ref T t)
+{
+    return t;
+}
+```
+The function `func()` does not destruct `s`, that obligation is transferred to
+the caller. In this way functions can be "pipelined" together, with the EMO being
+moved from one to the next by merely passing along the pointer.
+
+
+```
+S func(return S s)
+{
+    S s2;
+    return s2;  // error, can't return local by Move Ref
+}
+```
+
+### Returning an EMO by `ref`
+
+Returning a local EMO by `ref` is an error, just as for non-EMO objects.
+
+```
+ref S f(S s)
+{
+    return s; // error
+}
+
+ref S g()
+{
+    S s;
+    return s; // error
+}
+```
+
+### Passing an EMO by `ref`
+
+The semantics of passing an EMO by `ref` are the same as for non-EMO objects.
+For example,
+
+```
+void func(ref S);
+...
+S s;
+void func(s);
+```
+is allowed and the responsiblity of destructing `s` remains with the caller.
+
+
+### Assigning a `ref` EMO to a non-`ref` EMO
+
+Consider:
+```
+void func(ref S s)
+{
+    S t;
+    t = s;  // `s` is copied, not moved, to `t`
+}
+```
+This is because explicit `ref` to EMOs do not own the object they refer
+to, so the only choice is to copy it.
+
+The same semantics apply to passing a `ref` EMO to a non-`ref` parameter:
+```
+void func(S);
+
+void g(ref S s)
+{
+    func(s); // a copy of `s` is made
+}
+```
+
+### Forwarding
+
+Referring back to our desired forwarding function `fwd`:
+
+```
+ref S fwd(return ref S s) { return s; }
+```
+which we would like to be used to forward its argument like so:
+
+```
+void f(S s);
+...
+S s;
+f(fwd(s));
+f(fwd(S());
+```
+Ownership of the argument to fwd() is retained by the caller,
+and so the caller will be responsible for its destruction.
+When the call is made to f(), a copy is made.
+
+If `fwd()` is inlineable, however, the compiler is allowed to examine
+its implementation. If implementation actually does return only a ref to the
+parameter `s`, the copy can be elided if it was the last use of `s`.
+Much like NRVO, it is implementation
+dependent whether this copy elision takes place.
+
+
+
+### EMO and Garbage Collection
+
+The current semantics are defined so that a naive compacting garbage collector
+can move objects with a mere memcpy. The restriction is that no fields of an object
+can contain a pointer to that same object. Currently, there isn't a naive compacting
+collector for D, so this hasn't been a problem. A collector that tracks where the
+pointers are in an object would not have this issue.
+
+Ironically, because move constructors execute arbitrary code, this likely would interfere
+with the collector keeping track of the state of memory during a collection cycle,
+meaning that objects with move constructors would likely be pinned (treated as immoveable).
+
+
+### Class Objects
+
+Class objects cannot have move constructors or move assignment operators.
```

Comment:
Please expand upon this, with reasoning here. If it has something to do with being reference type, then mention said reason.